internal/events/eventer.go

```diff
@@ -196,25 +190,33 @@ func Setup(ctx context.Context, cfg *config.EventConfig) (*Eventer, error) {
 			// driver close
 			cl()
 		},
-		metrics: &metricInstruments,
+		msgInstruments: metricInstruments,
 	}, nil
 }
 
-func recordMetrics(m messageInstruments) func(h message.HandlerFunc) message.HandlerFunc {
+func recordMetrics(instruments *messageInstruments) func(h message.HandlerFunc) message.HandlerFunc {
 	metricsFunc := func(h message.HandlerFunc) message.HandlerFunc {
-		return func(message *message.Message) ([]*message.Message, error) {
-			producedMessages, err := h(message)
-
-			for _, msg := range producedMessages {
-				if publishedAt := msg.Metadata.Get(PublishedKey); publishedAt != "" {
-					if parsedTime, err := time.Parse(time.RFC3339, publishedAt); err == nil {
-						processingTime := time.Since(parsedTime)
-						m.messageProcessingTimeHistogram.Record(msg.Context(), processingTime.Milliseconds())
-					}
+		return func(msg *message.Message) ([]*message.Message, error) {
+			var processingTime time.Duration
+			if publishedAt := msg.Metadata.Get(PublishedKey); publishedAt != "" {
+				if parsedTime, err := time.Parse(time.RFC3339, publishedAt); err == nil {
+					processingTime = time.Since(parsedTime)
 				}
 			}
 
-			return producedMessages, err
+			// Defer the DLQ tracking logic to after the message has been processed by other middlewares,
+			// including the deferred PoisonQueue middleware functionality,
+			// so that we can check if it has been poisoned or not.
+			defer func() {
```

Comment:
Why use `defer` here, rather than capturing the returned values?

It works, but it feels somewhat more magical than simply:

```golang
res, err := h(msg)

isPoison := msg.MetadataGet(middleware.ReasonForPoisonedKey) != ""
instruments.messageProcessingTimeHistogram.Record(
    msg.Context(),
    processingTime.Milliseconds(),
    metric.WithAttributes(attribute.Bool("poison", isPoisoned)),
)

return res, err
```

Reply:
Oh, I merged it the moment this comment appeared. 

Nice suggestion, I'll follow up with another PR.

Reply:
^

https://github.com/stacklok/minder/pull/2069